feat: improvements to esbuild bundling

eduardoboucas opened the pull request:
**- Summary**

This PR includes [a bypass](https://github.com/netlify/zip-it-and-ship-it/blob/121bb49456a35ab99139160f3b99e12db1002d0f/src/main.js#L122-L124) for the legacy bundling logic when using esbuild, removing unnecessary computations and thus improving performance.

It also adds support for TypeScript.

Finally, to comply with the `max-lines` ESLint rule, it breaks the Go, JS and Rust runtime-specific logic into different files.

I'm sorry that the PR is larger than what I'd like, but the three changes are heavily interconnected.

**- Description for the changelog**

feat: improvements to esbuild bundling

**- A picture of a cute animal (not mandatory but encouraged)**

![maxresdefault](https://user-images.githubusercontent.com/4162329/107518743-45d17f00-6ba7-11eb-8ded-00f9e79b2fcb.jpg)

ehmicky commented on src/info.js:
```diff
@@ -37,7 +37,7 @@ const getFunctionInfo = async function (srcPath) {
   }
 
-  if (extension === '.zip' || extension === '.js') {
+  if (['.js', '.ts', '.zip'].includes(extension)) {
```
Should we add `.cjs` and `.mjs`?
There is also the older / less common `.es6`.

eduardoboucas replied:
To revisit once we add support for additional extensions.

ehmicky commented on src/main.js:
```diff
@@ -169,8 +117,12 @@ const getListedFunctionFiles = async function (
 
-const getSrcFiles = function ({ runtime, stat, mainFile, extension, srcPath, srcDir, pluginsModulesPath }) {
+const getSrcFiles = function ({ runtime, stat, mainFile, extension, srcPath, srcDir, pluginsModulesPath, useEsbuild }) {
   if (runtime === 'js' && extension === '.js') {
```
Should this be changed to `runtime === 'js' && extension !== '.zip'`? So it works with `.ts`.

eduardoboucas replied:
As per our chat, I'm temporarily reverting TypeScript support until we're comfortable with how to proceed.

ehmicky commented on src/main.js:
```diff
@@ -70,66 +78,6 @@ const zipFunction = async function (
-  return destPath
-}
-
-const RUNTIMES = {
```
Nice splitting into smaller files! :+1:

ehmicky commented on src/info.js:
```diff
@@ -37,7 +37,7 @@ const getFunctionInfo = async function (srcPath) {
-  if (extension === '.zip' || extension === '.js') {
+  if (['.js', '.ts', '.zip'].includes(extension)) {
     return { runtime: 'js', name, filename, stat, mainFile, extension, srcPath, srcDir }
   }
```
`getMainFile()` should also be fixed to support the new file extensions.

eduardoboucas replied:
As per our chat, I'm temporarily reverting TypeScript support until we're comfortable with how to proceed.

ehmicky commented on src/zip_node.js:
```diff
@@ -69,7 +70,7 @@ const zipNodeJsWithEsbuild = async function ({
     const { srcFile, stat } = await addStat(bundledFilePath)
 
-    addEntryFile(destFolder, archive, filename, bundledFilePath)
+    addEntryFile(destFolder, archive, jsFilename, bundledFilePath)
```
Should `addEntryFile()` only use `bundleFilePath` instead of `jsFilename`? Since this is the file actually being required.
This would also remove the need to fix the file extension, providing esbuild always outputs a `.js` file.

eduardoboucas replied:
`bundleFilePath` will contain a full path, whereas `jsFilename` contains the relative path, which is what we want to include in the entry file. Or am I misunderstanding?

eduardoboucas replied:
> This would also remove the need to fix the file extension, providing esbuild always outputs a .js file.

You actually need to provide an output file with a `.js` extension so that it builds for the right target. If the output file is `.ts`, it outputs TypeScript.

(There may be an option to modify this behaviour I don't know about.)

ehmicky replied:
Makes sense! :+1:

ehmicky commented on src/info.js:
```diff
@@ -37,7 +37,7 @@ const getFunctionInfo = async function (srcPath) {
     return {}
   }
 
-  if (extension === '.zip' || extension === '.js') {
```
Should we add some automated tests to cover the new file extension(s)?

eduardoboucas replied:
To revisit once we add support for new extensions.

eduardoboucas commented:
@ehmicky I've addressed your comments. Also, added support for the esbuild alpha environment variables in https://github.com/netlify/zip-it-and-ship-it/pull/323/commits/c819058bc09e0a6fb431410d47ed5410d5b05e6c.

ehmicky reviewed (approved):
:tada:
